### PR TITLE
Integrate new predict rerank endpoint

### PR DESCRIPTION
**Pull request by jotare**
### Description
Describe the proposed changes made in this PR.

### How was this PR tested?
Describe how you tested this PR.

### CODE REVIEW

File: nucliadb_models/src/nucliadb_models/search.py [hunk 1]
```diff
@@ -505,6 +521,11 @@ class SearchParamDefaults:
         title="Search features",
         description="List of search features to use. Each value corresponds to a lookup into on of the different indexes. `document`, `paragraph` and `vector` are deprecated, please use `fulltext`, `keyword` and `semantic` instead",  # noqa
     )
+    reranker = ParamDefault(
+        default=Reranker.MULTI_MATCH_BOOSTER,
+        title="Reranker",
+        description="Reranker let you specify which method you want to use to rerank your results at the end of retrieval",
```
Comment (lferran) [inline nucliadb_models/src/nucliadb_models/search.py:527]: Maybe you could extend the description with a short sentence about each rerank option?

File: nucliadb/src/nucliadb/search/search/query.py [hunk 2]
```diff
@@ -544,6 +560,29 @@ def get_max_tokens_answer(self) -> Optional[int]:
...
+    async def adjust_page_size(self, request: nodereader_pb2.SearchRequest, reranker: Reranker):
+        """Some rerankers want more results than the requested by the user so
+        reranking can have more choices. This function adjust the number of
+        retrieval results requested according to the reranker needs."""
+
+        if not reranker.needs_extra_results:
+            return
+
+        async with datamanagers.with_ro_transaction() as txn:
+            external_provider = await datamanagers.kb.get_external_index_provider_metadata(
```
Comment (lferran) [inline nucliadb/src/nucliadb/search/search/query.py:572]: If you only want to know if a kb has external index provider or not, I think you should hard cache this in memory as it is not modifiable per KB

File: nucliadb_models/src/nucliadb_models/search.py [hunk 3]
```diff
@@ -382,6 +382,22 @@ class SortOptions(BaseModel):
...
+    - Multi-match booster: given a set of results from different sources, e.g.,
+      keyword and semantic search boost results appearing in both sets
+
+    - Predict reranker: after retrieval, send the results to Predict API to
+      rerank it. This method uses a reranker model, so one can expect better
+      results at expenses of more latency
+
+    """
+
+    MULTI_MATCH_BOOSTER = ("multi_match_booster",)
```
Comment (lferran) [inline nucliadb_models/src/nucliadb_models/search.py:397]: why a tuple?

File: nucliadb/src/nucliadb/search/search/find_merge.py [hunk 4]
```diff
@@ -129,16 +129,17 @@ async def set_resource_metadata_value(
...
     kbid: str,
     show: list[ResourceProperties],
     field_type_filter: list[FieldTypeName],
     extracted: list[ExtractedDataTypeName],
     highlight: bool = False,
     ematches: Optional[list[str]] = None,
-):
+) -> tuple[dict[str, FindResource], list[str]]:
+    find_resources: dict[str, FindResource] = {}
+    best_matches: list[str] = []
```
Comment (lferran) [inline nucliadb/src/nucliadb/search/search/find_merge.py:141]: this is wrong no? we will not be fetching anything

File: nucliadb_models/src/nucliadb_models/search.py (hunk 1 above)
Comment (jotare) [reply nucliadb_models/src/nucliadb_models/search.py:527]: I extended the information in the `Rerank` enum, I thought it'd be better, although not sure how is the end result in the docs

File: nucliadb_models/src/nucliadb_models/search.py (hunk 3 above)
Comment (jotare) [reply nucliadb_models/src/nucliadb_models/search.py:397]: Oh, I guess I misplaced a `,` and `ruff` added the parens. Fixing it

File: nucliadb/src/nucliadb/search/search/query.py [hunk 5]
```diff
@@ -121,6 +126,7 @@ def __init__(
         rephrase_prompt: Optional[str] = None,
         max_tokens: Optional[MaxTokens] = None,
         hidden: Optional[bool] = None,
+        reranker: Optional[Reranker] = None,
```
[the remark this answers is not in the document]
Comment (jotare) [reply nucliadb/src/nucliadb/search/search/query.py:129]: No strong opinion, I'll change it

File: nucliadb/src/nucliadb/search/search/query.py (hunk 2 above)
Comment (jotare) [reply nucliadb/src/nucliadb/search/search/query.py]: Not sure if now is much better, as the first query for KBs with our index will use two transactions, but I guess it's fine

File: nucliadb/src/nucliadb/search/search/find.py [hunk 6]
```diff
@@ -278,6 +224,14 @@ async def _external_index_retrieval(
         max_parallel_operations=50,
     )
 
+    # once hydrated, we send the results to rerank
+    reranker = query_parser.reranker
+    await reranker.rerank_find_response(
+        kbid, query_parser.query, retrieval_results, query_parser.page_size
+    )
+    # XXX: done later, could be substituted though
+    retrieval_results.best_matches.clear()
```
Comment (javitonino) [inline nucliadb/src/nucliadb/search/search/find.py:233]: Is this even needed? It's replaced 4 lines below.

**Review by javitonino — APPROVED** [review]

File: nucliadb/src/nucliadb/search/search/find.py (hunk 6 above)
Comment (jotare) [reply nucliadb/src/nucliadb/search/search/find.py:233]: Removed :)

**Comment by jotare** [comment]
Changed to merge against dev: https://github.com/nuclia/nucliadb/pull/2485

**Pull request closed without merging**